spec/inputs/http_poller_spec.rb

```diff
@@ -2,87 +2,151 @@
 require 'logstash/inputs/http_poller'
 
 describe LogStash::Inputs::HTTP_Poller do
-  let(:settings) {
+  let(:queue) { Queue.new }
+  let(:default_interval) { 5 }
+  let(:default_name) { "url1 " }
+  let(:default_url) { "http://localhost:1827" }
+  let(:default_urls) {
     {
-      "interval" => 10,
-      "request_timeout" => 1,
-      "connect_timeout" => 1,
-      "socket_timeout" => 1,
-      "automatic_retries" =>1,
-      "type" => "something-fancy"
+      default_name => default_url
     }
   }
-  context "with json codec, when remote server is failing" do
-    it "should not add an event" do
-      subject = LogStash::Inputs::HTTP_Poller.new(settings.merge({"urls" => {"url1" => "http://mytesturl.mydomain"}}))
-      # no stub, and an incorrect domain, this should result in no response.
+  let(:default_opts) {
+    {
+      "interval" => default_interval,
+      "urls" => default_urls,
+      "codec" => "json"
+    }
+  }
+  let(:mod) { LogStash::Inputs::HTTP_Poller }
 
-      logstash_queue = Queue.new
-      subject.run_once logstash_queue
+  subject { mod.new(default_opts) }
+
+  describe "#run" do
+    it "should run at the specified interval" do
+      expect(Stud).to receive(:interval).with(default_interval).once
+      subject.send(:run, double("queue"))
+    end
+  end
 
-      expect(logstash_queue.size).to eq(0)
+  describe "#run_once" do
+    it "should issue an async request for each url" do
+      default_urls.each { |name, url|
+        expect(subject).to receive(:request_async).with(queue, name, url).once
+      }
 
-      subject.teardown
+      subject.send(:run_once, queue)
     end
+  end
 
-    it "should create an event with an error" do
-      subject = LogStash::Inputs::HTTP_Poller.new(settings.merge({"urls" => {"url1" => "http://mytesturl.mydomain"}}))
-      subject.client.stub("http://mytesturl.mydomain", body:"not found",  code: 404)
+  shared_examples("matching metadata") {
+    let(:metadata) { event["@metadata"]["http_poller"] }
 
-      logstash_queue = Queue.new
-      subject.run_once logstash_queue
-      event = logstash_queue.pop
+    it "should have the correct name" do
+      expect(metadata["name"]).to eql(name)
+    end
+
+    it "should have the correct url" do
+      expect(metadata["url"]).to eql(url)
+    end
 
-      expect(event["responseCode"]).to eq(404)
-      expect(event["success"]).to eq(false)
-      subject.teardown
+    it "should have the correct code" do
+      expect(metadata["code"]).to eql(code)
     end
+  }
+
+  shared_examples "unprocessable_requests" do
+    let(:poller) {
+      p = LogStash::Inputs::HTTP_Poller.new(settings)
+      allow(p).to receive(:handle_failure).and_call_original
+      allow(p).to receive(:handle_success)
+      p
+    }
+    subject(:event) {
+      poller.send(:run_once, queue)
```

Comment:
similar to my `allow` comment; it feels weird having let/subject blocks _do_ things, rather than declare things. I admit in some cases it's necessary, but my preference is to lean towards having before blocks _do_ things and let/subject blocks simply declare them